clients/src/main/java/org/apache/kafka/common/security/authenticator/DefaultKafkaPrincipalBuilder.java

```diff
@@ -118,8 +122,14 @@ public KafkaPrincipal build(AuthenticationContext context) {
             SaslServer saslServer = ((SaslAuthenticationContext) context).server();
             if (SaslConfigs.GSSAPI_MECHANISM.equals(saslServer.getMechanismName()))
                 return applyKerberosShortNamer(saslServer.getAuthorizationID());
-            else
+            else {
+                Object principalType = saslServer.getNegotiatedProperty(PRINCIPAL_TYPE);
+                if(principalType != null){
```

Comment:
When I read the code, I don't understand the logic, I think we need to comments it either here or the design doc. If principal type is missing, are we using anonymous ones?

Reply:
Old way of using anonymous should be rejected.